src/test/java/com/pngencoder/PngEncoderTest.java

```diff
@@ -297,9 +329,10 @@ private static int[] toIntArgb(BufferedImage bufferedImage) {
         int height = bufferedImage.getHeight();
         int[] argbImage = new int[width * height];
 
+        int writePtr = 0;
         for (int y = 0; y < height; y++) {
             for (int x = 0; x < width; x++) {
-                argbImage[y * width + x] = bufferedImage.getRGB(x, y);
+                argbImage[writePtr++] = bufferedImage.getRGB(x, y);
```

Comment:
Ok, at that place performance does not matter. You can usually not expect the JVM to be smart enough to move the subexpression `y*width` out of the x loop.  So it will be very likely faster. 

But as said here the performance does not matter. I'll change that back.

Reply:
When in doubt: measure! I'd say the JVM is smart enough to inline both of these versions enough so that `getRGB` is the bottleneck.

182dfd1
I've tested two loops using `Random.nextInt()` and  `BufferedImage.getRGB`. For these, the two solutions are pretty much equal. 

b14f4af
If I remove the "work" and just measure loop performance, the difference is greater.

I guess the pointer version can be inlined better, but it doesn't matter when there is significant other work to be done (e.g. `getRGB`)

https://github.com/pngencoder/pngencoder/pull/40

Reply:
I would argue that the less variables you access the higher is the chance that the variables are not spilled on the stack and can be hold directly in a register. Stack pages should always be in the CPU cache, but if not you get a real performance hit. Also if you do [Common Subexpression Elimination](https://en.wikipedia.org/wiki/Common_subexpression_elimination) yourself, you don't need to trust the compiler to be smart enough to do this for you. The compiler usually should perform this optimizations but sometimes he doesn't. E.g. if the register pressure is too high he might think that rematerialzation of some values might be better than caching them on the stack. He might be right, but this is not always the case.  

Independent of the language (i.e. it's the same for C++ etc.) for performance critical stuff you should not trust the compiler to be smart. Stuff like the [Vector API](https://openjdk.org/jeps/426) gets developed as smart compiler optimizations are likely to break sometimes with other changes. 

This really only matters if you care about low level performance. I.e. inner loops. For e.g. the test driver here it really does not matter. Also for most other code the JVM optimizations are usually more then enough.